cprover_bindings/src/irep/goto_binary_serde.rs

```diff
@@ -0,0 +1,1339 @@
+    fn nof_named_sub(&self, cache: &IrepNumbering) -> usize {
+        cache.inv_cache.keys[self.start_index + self.nof_sub(cache) + 2]
+    }
+
+    /// Returns a pair formed of a NumberedString and NumberedIrep describing the named
+    /// sub number `i` of this numbered irep.
+    fn named_sub(
+        &self,
+        cache: &IrepNumbering,
+        named_sub_idx: usize,
+    ) -> (NumberedString, NumberedIrep) {
+        let start_index = self.start_index + self.nof_sub(cache) + 2 * named_sub_idx + 3;
+        (
+            cache.inv_string_cache[cache.inv_cache.keys[start_index]],
+            cache.inv_cache.index[cache.inv_cache.keys[start_index + 1]],
+        )
+    }
+}
+
+/// GOTO binary serializer.
+struct GotoBinarySerializer<'a, W>
+where
+    W: Write,
+{
+    writer: &'a mut W,
+
+    /// In-memory temporary buffer, contents get flushed after each object
+    buf: Vec<u8>,
+
+    /// Numbering used for structural sharing.
+    numbering: IrepNumbering,
+
+    /// Counts how many times a given irep was written.
+    irep_count: Vec<usize>,
+
+    /// Counts how many times a given string was written.
+    string_count: Vec<usize>,
+}
+
+impl<'a, W> GotoBinarySerializer<'a, W>
+where
+    W: Write,
+{
+    /// Constructor.
+    fn new(writer: &'a mut W) -> Self {
+        GotoBinarySerializer {
+            writer,
+            buf: Vec::new(),
+            numbering: IrepNumbering::new(),
+            irep_count: Vec::new(),
+            string_count: Vec::new(),
+        }
+    }
+
+    /// Returns statistics about the serializer.
+    fn get_stats(&self) -> GotoBinarySerializerStats {
+        GotoBinarySerializerStats::new(self)
+    }
+
+    /// Adds an InternedString uid to the "written" cache, returns true iff was never written before.
+    fn is_first_write_string(&mut self, u: usize) -> bool {
+        if u >= self.string_count.len() {
+            self.string_count.resize(u + 1, 0);
+        }
+        let count = self.string_count[u];
+        self.string_count[u] = count.checked_add(1).unwrap();
+        count == 0
+    }
+
+    /// Adds an Irep uid to the "written" cache, returns true iff it was never written before.
+    fn is_first_write_irep(&mut self, u: usize) -> bool {
+        if u >= self.irep_count.len() {
+            self.irep_count.resize(u + 1, 0);
+        }
+        let count = self.irep_count[u];
+        self.irep_count[u] = count.checked_add(1).unwrap();
+        count == 0
+    }
+
+    /// Flushes the temporary buffer to the external writer,
+    /// flushes the writer and clears the temporary buffer.
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.write_all(&self.buf)?;
+        self.buf.clear();
+        Ok(())
+    }
+
+    /// Writes a single byte to the temporary buffer.
+    fn write_u8(&mut self, u: u8) -> io::Result<()> {
+        self.buf.push(u);
+        Ok(())
+    }
+
+    /// Writes a usize to the temporary buffer using 7-bit variable length encoding.
```

Comment:
Please expand on the 7-bit variable length encoding, e.g. by mentioning that the 8th bit indicates whether there are more bytes or not.

Reply:
fixed